src/java/org/apache/cassandra/db/compaction/UnifiedCompactionContainer.java

```diff
@@ -141,6 +143,14 @@ public CompactionStrategyContainer reload(@Nonnull CompactionStrategyContainer p
         return create(previous, factory, compactionParams, reason, enableAutoCompaction);
     }
 
+    @Override
+    public boolean shouldReload(CompactionParams params, ReloadReason reason)
+    {
+        return reason != CompactionStrategyContainer.ReloadReason.METADATA_CHANGE
+               || !params.equals(getMetadataCompactionParams())
+               || hasVector != factory.getRealm().metadata().hasVectorType();
```

Comment:
Concurrent modification of the metadata looks fine here: if the metadata changes between this call and the time `fromOptions` gets it, we will apply the settings change early.